### ISSUE TITLE
How to pause and resume a slow @KafkaListener?  

### ISSUE THREAD

**Issue by dusan-eremic** [spring-projects/spring-kafka#546]
Hello,

I see that this feature is announced here... https://spring.io/blog/2016/06/07/spring-for-apache-kafka-1-0-release-candidate-1-available

...but after following the link to the ref docs https://docs.spring.io/spring-kafka/docs/1.0.0.RC1/reference/htmlsingle I can't find more details how to use this feature.

Can anyone help with sharing up-to-date documentation or a code snippet demonstrating pause/resume feature? 

Thanks

**Comment by artembilan** [comment]
The `@KafkaKistener` method can be supplied with the `KafkaConsumer` argument.
And exactly that one you can use to `pause` and `resume`.

**Comment by garyrussell** [comment]
Please upgrade to 1.3.3 or higher - the complex algorithm needed to pause/resume to prevent a rebalance is no longer needed (Since KIP-62).

2.0 or higher gives listeners access to the `Consumer` object.

See the [current reference manual](https://docs.spring.io/spring-kafka/reference/html/) the 1.0.0.RC1 is very, very old.

**Comment by dusan-eremic** [comment]
@garyrussell and @artembilan My bad, thank you sharing the right docs. 

What is the right way to get the list of all partitions the consumer is subscribed to? Can I use `assignment`?

**Comment by garyrussell** [comment]
No; the pause will take effect only after the currently polled records are drained; there is no mechanism to find out how many records remain.

You could throw an exception after pausing and configure the container with a `SeekToCurrentErrorHandler` which will discard the remaining records and they will be re-fetched on the next poll after resuming.

I generally prefer to always set `enable.auto.commit` to false and let the container handle the commits.

**Comment by dusan-eremic** [comment]
Thank you Gary.

Would `max.poll.records=1` ensure that after calling pause there is no more records waiting in the buffer?

**Comment by artembilan** [comment]
That’s true. As far as your listener is a single thread and you call `pause()` before returning to the container to let it call the next `poll()`. Well, I even would say that it doesn’t matter what the value is for `max.poll.records` if process current records properly in between latest `poll()` and `pause()`.

**Comment by artembilan** [comment]
This issue is addressed via: https://github.com/spring-projects/spring-kafka/commit/15400928ff725852f3d3f667946938c089d9391c. Now the `ListenerContainer` exposes `pause()/resume()` functionality which can be used in any target code.

**Comment by V-Rajasekar** [comment]
Can you please share a code snippet how to invoke the pause( )/resume( ) on the listener instead on a consumer as advised in spring-kafka reference doc as it is not always thread-safe.

I tried injecting or Autowiring the listener in the kafka-consumer but it lead to below cyclic error  

The dependencies of some of the beans in the application context form a cycle: because 
┌─────┐
|  myListener defined in class path resource [no/demo/services/configuration/KafkaConfiguration.class]
↑     ↓
|  myMessageConsumer (field org.springframework.kafka.listener.KafkaMessageListenerContainer no.demo.services.MyMessageConsumer.myListener)
└─────┘

**Comment by garyrussell** [comment]
Please don't ask the same question in multiple places. I will add a snippet to your https://github.com/spring-projects/spring-kafka/issues/650

**Comment by SpiralA6** [comment]
In the javadocs for **GenericMessageListener** it mentions that two (out of three) of the methods support

    _Invoked with null data when a poll returns no data (enabling resume)._ 

Should this apply to all methods?
Also, this does not appear to be the case looking at the code in **KafkaMessageListenerContainer**

```
					if (**records != null && records.count() > 0**) {
						if (this.containerProperties.getIdleEventInterval() != null) {
							lastReceive = System.currentTimeMillis();
						}
						**invokeListener**(records);
					}
```

Am I missing something?

**Comment by SpiralA6** [comment]
Please ignore. I noticed this is closed so will start a new ticket.
